lib/ecto/query/planner.ex

```diff
@@ -593,6 +594,23 @@ defmodule Ecto.Query.Planner do
     end
   end
 
+  defp merge_cache(:with_cte, _query, nil, cache_and_params, _operation, _adapter) do
+    cache_and_params
+  end
+
+  defp merge_cache(:with_cte, _query, with_expr, cache_and_params, operation, adapter) do
+    fun = &{&3, merge_cache(&1, &2, &3, &4, operation, adapter)}
+
+    Enum.reduce with_expr.queries, cache_and_params, fn
+      {_, %Ecto.Query{} = query}, acc ->
+        {_, acc} = traverse_exprs(query, operation, acc, fun)
+        acc
+
+      _, acc ->
+        acc
```

Comment:
In this branch we need to call `cast_and_merge_params`, like we do for other QueryExpr here: https://github.com/elixir-ecto/ecto/blob/3680d9b408672154c0ca16327b47ef42a89a2a8b/lib/ecto/query/planner.ex#L544-L556

Reply:
I have commented on the planner_test file about a test we should add that will expose the need for this.